### ISSUE TITLE
Code Base Merge Challenge 2018™

### ISSUE THREAD

**Issue by TurtleP** [TurtleP/LovePotion#39]
I'm sure both @NotQuiteApex and myself have (briefly) discussed this before:

There should be an attempt to keep the code bases on the same page. I've thought about this for quite some time and I wasn't sure how to go about it, mainly because I have to see what needs changing to make it work on each platform. I definitely want to avoid a bunch of `#ifdef` stuff clogging up the code, but it might be necessary for some things.

1. Using module wrappers. Things like the `system` module clearly have differences between Switch and 3DS and there should be something to keep them apart, but still allowing the code to be built for either platform.

2. We could do a `git diff` and apply patches to whichever version (most likely when building 3DS from Switch, as it's more feature-complete), but that might not be a good option. This is because we'd wind up removing most of the Switch functionality that the 3DS diff would remove. like canvases or threads unless we can somehow allow the diff to fill in these missing pieces. This sounds like a pain to do.

Any other ideas are welcome and I'll discuss with others who might be more knowledgeable with regards to this.

tl;dr
- Merge common code so we don't have several branches just for the two devices. 
  - Except for feature requests and other enhancements
- **General improvements to enhance the user experience and overall system stability**

**Comment by TSnake41** [comment]
Update from 2019, I found other(s) way(s) to merge both code bases.
However, none of these are perfect so...
![merge challenge 2019-1](https://user-images.githubusercontent.com/8642647/51636299-33049700-1f59-11e9-93f7-deaae4e14ff1.jpg)
![merge challenge 2019-2](https://user-images.githubusercontent.com/8642647/51636300-33049700-1f59-11e9-8a34-ab9458a5178c.jpg)
(yes, it is handwritten :P)

**Comment by TurtleP** [comment]
To Do:

- [ ] Get 3DS compilation working
  - [ ] Use SDL_Color instead of a custom Color struct
  - [ ] Make Quads use Viewport struct, apply them in the Image class to a subtex when rendering
  - [ ] Switch to using SDL_Event (I think it works? I'll double check with WinterMute/fincs/etc) for Event::Poll
- [ ] Make boot.lua work regardless of platform (don't use double screens for 3DS)

**Comment by TurtleP** [comment]
Alright closing as I'm in the middle of the code merge. Stuff's going super smoothly for 3DS stuff! :)
